pkg/operator/controllers/machineset/machineset_controller.go

```diff
@@ -0,0 +1,100 @@
+package machineset
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	machinev1beta1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	maoclient "github.com/openshift/machine-api-operator/pkg/generated/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+)
+
+type Reconciler struct {
+	log *logrus.Entry
+
+	maocli maoclient.Interface
+	arocli aroclient.Interface
+}
+
+// MachineSet reconciler watches MachineSet objects for changes, evaluates total worker replica count, and reverts changes if needed.
+func NewReconciler(log *logrus.Entry, maocli maoclient.Interface, arocli aroclient.Interface) *Reconciler {
+	return &Reconciler{
+		log:    log,
+		maocli: maocli,
+		arocli: arocli,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.Features.ReconcileMachineSet {
+		return reconcile.Result{}, nil
+	}
+
+	machinesets, err := r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).List(ctx, metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=worker"})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	aroMachineset, err := r.maocli.MachineV1beta1().MachineSets(machineSetsNamespace).Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Count amount of total current worker replicas (including custom MachineSets, if present)
+	replicaCount := 0
+	for _, machineset := range machinesets.Items {
+		if machineset.Spec.Replicas != nil {
+			replicaCount += int(*machineset.Spec.Replicas)
+		}
+	}
+
+	// Verify that the MachineSet that was modified matches the cluster's InfraID, otherwise ignore it
+	matches, err := regexp.Match(instance.Spec.InfraID, []byte(aroMachineset.ObjectMeta.Name))
+	if err != nil {
+		r.log.Error(err)
+	}
+
+	if replicaCount < 3 && matches {
+		r.log.Info("Found less than 3 worker replicas. The MachineSet controller will attempt scaling.")
+		aroMachineset.Spec.Replicas = to.Int32Ptr(int32(1) + *aroMachineset.Spec.Replicas) // Add one replica to the object, and call Update
```

Comment:
Why are we just adding one replica, instead of adding `(3 - replicaCount)`?

Reply:
(Just a question, not necessarily a requested change.)

Reply:
I think your suggestion is better. With just incrementing 1, it has to reconcile twice in order to get back to 3, but using `3-replicaCount` cuts it down to a single reconcile.

I am pretty sure we'd only see this in a cluster without AZs where there's only one worker MachineSet, and the user attempts to scale down from 3 replicas to 1.